### ISSUE TITLE
Forward Pathing

### ISSUE THREAD

**Issue by exodist** [Test-More/Test2-Suite#93]
Back when I was first creating Test2::Suite I had a vision for how it could move forward when breaking changes were needed. I mostly forgot, and even started to move away from that path... I think I need to return to the original plan, and I will document it here as best I can.

With Test2::Suite tools, plugins, and bundles are all treated as disposable. This is why tools and plugins are separated out by functionality, and are very small units. With exception of Test2::Tools::Compare nearly everything in Test2::Tools::\* is very tiny.

The idea was that if Test2::Tools::Foo turned out to have serious problems, or be a bad idea, we could replace it by writing Test2::Tools::Bar and recommend people stop using ::Foo. The consequences are minimal, people already using ::Foo don't break, and people writing new tests can use ::Bar instead.

Bundles are also supposed to work this way, if a bundle includes a bunch of bad tools then simply write a new bundle, it should be easy since bundles have minimal logic and simply re-export tools and plugins.

Right now there are 2 modules in particular that we are having trouble with:
- Test2::Tools::Compare
- Test2::Bundle::Extended

The reason we are having trouble with Test2::Tools::Compare is because it really did break away from the small and disposable model. The extended bundle has a problem because the compare tools have a problem, but also because it is the "recommended" bundle, so replacing it is strange.

---

One of the things I as working on to try to address moving forward was the use of '+v#' pins. Another idea that is similar, but less crazy and more maintainable is the use of Foo::V# variants, like what perl5i uses.

I have a problem with this system. This works with perl5i because it is a single module the consumer uses. Test2::Suite is a ton of modules, and the consumer can use any number/combination of them. Having V variants of each requires a lot of overhead for the consumers. Even if we make it so that they all get a new variant at once with the same number it is still a lot of cognitive overhead.

I don't like either system here for most things, though I do still want to adopt the perl5i ::V# system for one specific thing: Recommended bundle.

I think we should create Test2::Bundle::V and Test2::Bundle::V1. Stop reommending 'Extended' and start recommending Test2::Bundle::V.  Loading bundle V will load the latest V# variant, but warn that the user should use V# (whatever is latest) instead. If someone uses V1 and we create V2 they will not break because we will not change V1 in a breaking way once it is created.

---

What about Test2::Tools::Compare?

Boils down to this: I think that including the compare functions `is`, `like`, etc and the dsl in a single tools module is/was a mistake.
- The dsl functions should be split into a new Tools::\* module (maybe Test2::Tools::Compare::DSL).
- We should deprecate (via a warning) the import of the dsl functions from Test2::Tools::Compare (but not break them!)
- The new DSL tools should be written in the new styles have been discussing, breaking compatability here is fine because it is a new package, people have to choose to switch
- The V1 bundle will export the new tools, the 'Extended' bundle will continue to provide the old.

---

What about is() and implicit end?

The more I think about it, the more I think not having an implicit end() was a bug, plain and simple. I suspect I was the only one who knew, and intentionally used the lack of implicit end as a feature. So lets just fix the damn bug. Things that break from fixing this bug break because a bug was letting them pass when they should not have. And the fix is easy, add an `etc` call to the structure.

That said, I also think we should give extra information when a test fails due to implicit end() to explain the situation better (this previously passed, but fails now, here is why....)

Since this is the only change to the non-dsl tools in Test2::Tools::Compare we will not need to replace Test2::Tools::Compare or create and variants for it.

---

So to summarize:
- Fix the implicit end bug
- Write a DSL tool module with our new DSL ideas (still in the works)
- Deprecate the dsl tools in Test2::Tools::Compare
- Recommended bundle is now Test2::Bundle::V and numeric variants as we need to make breaking changes.

**Comment by schwern** [comment]
> I think we should create Test2::Bundle::V and Test2::Bundle::V1. Stop reommending 'Extended' and start recommending Test2::Bundle::V. Loading bundle V will load the latest V# variant, but warn that the user should use V# (whatever is latest) instead. If someone uses V1 and we create V2 they will not break because we will not change V1 in a breaking way once it is created.

This could also solve a UI problem. Right now, as user, it feels like I have to assemble my own library from the myriad of pieces and parts and versions. Instead I just use Test2::Bundle::Extended without really understanding what's in it. I still have to go find the actual docs for individual pieces all over the place.

I think the Bundles should be the user facing module complete with full documentation of what is provided (we may be able to avoid duplication with Pod::Weaver). I'd go so far as to drop the "Bundle" part and call this Test2::V or something that the user will clearly recognize as "this is what you want to use when you use Test2".

> I think that including the compare functions is, like, etc and the dsl in a single tools module is/was a mistake.

Now that you've pointed it out, I agree. I don't know about putting all the DSL stuff into one module is the correct thing to do, we'll need to do some prototyping to see how that splits up.

**Comment by exodist** [comment]
Extended was dropped in favor of V1. Many of the other ideas from this ticket are not happening, closing now. Any remaining ideas in here that anyone wants to promote please open a new ticket.
